### ISSUE TITLE
gcloud preview app run ./app.yaml throwing error

### ISSUE THREAD

**Issue by redwood-egarcea** [GoogleCloudPlatform/nodejs-getting-started#12]
Seems to be caused by an unknown RuntimeError. 
```
RuntimeError: Unknown runtime 'nodejs'; supported runtimes are 'custom', 'go', 'java', 'java7', 'php', 'php55', 'python', 'python27', 'vm'.
```
Runs fine when executing on google.
```
gcloud preview app deploy app.yaml --set-default 
```

```
➜  1-hello-world git:(1-hello-world)  gcloud --verbosity debug preview app run ./app.yaml
DEBUG: Running gcloud.preview.app.run with Namespace(__calliope_internal_deepest_parser=ArgumentParser(prog='gcloud.preview.app.run', usage=None, description="*(BETA)* This command is used to run one or more of your modules in the local\ndevelopment application server.  This allows you to test and debug your app\nbefore deploying.  As an input it takes one or more ``RUNNABLES'' that should\nbe run locally.  A ``RUNNABLE'' can be a module's .yaml file, or a directory.\nIf given a directory, all modules found will be run.  You can run multiple\nmodules at the same time, even if they are implemented in different languages.", version=None, formatter_class=<class 'argparse.HelpFormatter'>, conflict_handler='error', add_help=False), admin_host=None, allow_skipped_files=False, api_host=None, appidentity_email_address=None, appidentity_private_key_path=None, auth_domain='gmail.com', blobstore_path=None, clear_datastore=False, cmd_func=<bound method Command.Run of <googlecloudsdk.calliope.backend.Command object at 0x10d869690>>, command_path=['gcloud', 'preview', 'app', 'run'], custom_entrypoint=None, datastore_consistency_policy='time', datastore_path=None, default_gcs_bucket_name=None, document=None, enable_cloud_datastore=False, enable_mvm_logs=False, enable_sendmail=False, format=None, h=None, help=None, host=None, http_timeout=None, jvm_flag=None, log_http=None, log_level=None, logs_path=None, markdown=None, max_module_instances=None, php_executable_path=None, project=None, python_startup_script=None, quiet=None, require_indexes=False, runnables=['./app.yaml'], show_mail_body=False, smtp_allow_tls=False, smtp_host=None, smtp_password=None, smtp_user=None, storage_path=None, trace_token=None, use_mtime_file_watcher=False, user_output_enabled=None, verbosity='debug', version=None).
DEBUG: Found Cloud SDK root: /Users/elliottgarcea/google-cloud-sdk
DEBUG: Found App Engine SDK root: /Users/elliottgarcea/google-cloud-sdk/platform/google_appengine
Module [default] found in file [/Users/elliottgarcea/Workspace/1-hello-world/app.yaml]
INFO: Looking for the Dockerfile in /Users/elliottgarcea/Workspace/1-hello-world
INFO: Using Dockerfile found in /Users/elliottgarcea/Workspace/1-hello-world
DEBUG: Running [dev_appserver.py] with: --allow_skipped_files=False --application=client-server-test --auth_domain=gmail.com --clear_datastore=False --datastore_consistency_policy=time --dev_appserver_log_level=debug --enable_cloud_datastore=False --enable_mvm_logs=False --enable_sendmail=False --log_level=debug --require_indexes=False --show_mail_body=False --skip_sdk_update_check=True --smtp_allow_tls=False --use_mtime_file_watcher=False /Users/elliottgarcea/Workspace/1-hello-world/app.yaml
INFO     2015-08-07 20:37:48,516 application_configuration.py:399] No version specified. Generated version id: 20150807t203748
INFO     2015-08-07 20:37:48,516 devappserver2.py:763] Skipping SDK update check.
DEBUG    2015-08-07 20:37:48,560 app_identity_stub.py:197] Successfully loaded Application Default Credentials.
INFO     2015-08-07 20:37:48,567 api_server.py:205] Starting API server at: http://localhost:61437
INFO     2015-08-07 20:37:48,568 api_server.py:646] Applying all pending transactions and saving the datastore
INFO     2015-08-07 20:37:48,568 api_server.py:649] Saving search indexes
Traceback (most recent call last):
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/dev_appserver.py", line 83, in <module>
    _run_file(__file__, globals())
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/dev_appserver.py", line 79, in _run_file
    execfile(_PATHS.script_file(script_name), globals_)
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/devappserver2.py", line 1033, in <module>
    main()
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/devappserver2.py", line 1026, in main
    dev_server.start(options)
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/devappserver2.py", line 818, in start
    self._dispatcher.start(options.api_host, apis.port, request_data)
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/dispatcher.py", line 193, in start
    _module, port = self._create_module(module_configuration, port)
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/dispatcher.py", line 278, in _create_module
    threadsafe_override=threadsafe_override)
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/module.py", line 1536, in __init__
    super(ManualScalingModule, self).__init__(**kwargs)
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/module.py", line 555, in __init__
    self._module_configuration)
  File "/Users/elliottgarcea/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/module.py", line 251, in _create_instance_factory
    sorted(repr(k) for k in self._RUNTIME_INSTANCE_FACTORIES))))
RuntimeError: Unknown runtime 'nodejs'; supported runtimes are 'custom', 'go', 'java', 'java7', 'php', 'php55', 'python', 'python27', 'vm'.
DEBUG: Error [1] while running DevAppSever with: [--allow_skipped_files=False --application=client-server-test --auth_domain=gmail.com --clear_datastore=False --datastore_consistency_policy=time --dev_appserver_log_level=debug --enable_cloud_datastore=False --enable_mvm_logs=False --enable_sendmail=False --log_level=debug --require_indexes=False --show_mail_body=False --skip_sdk_update_check=True --smtp_allow_tls=False --use_mtime_file_watcher=False /Users/elliottgarcea/Workspace/1-hello-world/app.yaml]
DEBUG: (gcloud.preview.app.run) DevAppSever failed with error code [1]
Traceback (most recent call last):
  File "/Users/elliottgarcea/google-cloud-sdk/./lib/googlecloudsdk/calliope/cli.py", line 548, in Execute
    result = args.cmd_func(cli=self, args=args)
  File "/Users/elliottgarcea/google-cloud-sdk/./lib/googlecloudsdk/calliope/backend.py", line 1102, in Run
    result = command_instance.Run(args)
  File "/Users/elliottgarcea/google-cloud-sdk/./lib/googlecloudsdk/appengine/app_commands/run.py", line 334, in Run
    runner.Start(*runnables)
  File "/Users/elliottgarcea/google-cloud-sdk/./lib/googlecloudsdk/appengine/lib/dev_appserver_adapter.py", line 138, in Start
    raise DevappserverExecutionError(return_code, argv)
DevappserverExecutionError: DevAppSever failed with error code [1]
ERROR: (gcloud.preview.app.run) DevAppSever failed with error code [1]
```

**Comment by jonparrott** [comment]
This is by design. ``gcloud preview app run`` is intended for using the sandbox and ``compat`` runtimes. Node.js is a new runtime. To run locally, you just need to execute ``npm start`` - just as you would while developing any other Node.js application.

**Comment by ejferg** [comment]
@jonparrott Only problem with that is you wouldn't be able to use appengine-nodejs locally.

**Comment by jonparrott** [comment]
We do not recommend using appengine-nodejs. We recommend using gcloud-node,
which works everywhere.

**Comment by ejferg** [comment]
@jonparrott I do plan on using using gcloud-node, but that wouldn't help me if I wanted to use the Memcache API or Task queue API for example. What would be your suggestion in those cases? Running Redis (is that as cost a effective as using the Memcache API). Just looking for recommendations.

**Comment by jonparrott** [comment]
Keep in mind that Managed VMs (and node.js support in GAE) is in beta and features are subject to change, and that achieving parity with non-managed VMs is a goal.

For Memcache:
* You can continue to use appengine-nodejs, and if we do introduce memcache support for MVMs, you can switch easily. The drawback is that you won't be able to easily test locally unless you add logic to switch to another caching library locally.
* You can use Redis. Redis can be [deployed to GCE](https://cloud.google.com/solutions/redis/click-to-deploy) or you can use managed services such as [Redis Cloud](https://redislabs.com/pricing) and [RedisToGo](http://redistogo.com/?language=en). Redis Cloud's free tier is more generous than RedisToGo's. Redis is one of my favorite databases for purposes beyond caching.

For Task Queues:
* Again, you can continue to use appengine-nodejs, but it may be more difficult to migrate, and even more difficult to test as task queues are a complex service.
* If you're using pull queues, you can try the [Alpha Task Queue REST API](https://cloud.google.com/appengine/docs/python/taskqueue/rest/).
* Depending on your use case, you can use [Cloud Pub/Sub](https://cloud.google.com/pubsub) for task queueing in more. In fact, this tutorial uses pub/sub precisely for that purpose in [this section](https://cloud.google.com/nodejs/getting-started/using-pub-sub).
* Use another queue. There seems to be quite a lot of options, such as celery support for node.js. RabbitMQ and other brokers can be easily deployed to GCE and used from anywhere. Also, you can use many of these queues across languages.

**Comment by ejferg** [comment]
@jonparrott Great! Reply. Yeah Redis is awesome. I also saw gcloud-kvstore referenced on the glcoud github page (not sure what the speed would be like??) for a key-value store. I'm going to be using Pub/Sub for other things in the application so it may make since to also use it for task queueing.

**Comment by jonparrott** [comment]
> gcloud-kvstore

It's based on Cloud Datastore so the latency will be about an order of magnitude higher than memcache, but it's a durable storage system.

**Comment by douglascorrea** [comment]
@jonparrott can I use runtime: nodejs in production for deploying managed VMs or should I use dockerfile and runtime: custom?

**Comment by jonparrott** [comment]
`runtime: nodejs` is the preferred way to run a node.js app on managed vms. It is considered beta (managed vms itself is in beta).

Behind the scenes, `runtime: nodejs` uses a Google-provided nodejs Docker image. If needed, you can use `gcloud preview app gen-config` which will create a Dockerfile for you that you can customize.

**Comment by Chipintoza** [comment]
@jonparrott thank you for your defination..
I use gcloud preview app gen-config and i want to add Redis to the Dockerfile:

before "npm start" when i add:

---------------------------------------------------------------------------------------------------------
# Install Redis.
RUN \
  cd /tmp && \
  (curl http://download.redis.io/releases/redis-stable.tar.gz | tar xvz) && \
  cd redis-stable && \
  make && \
  make install && \
  cp -f src/redis-sentinel /usr/local/bin && \
  mkdir -p /etc/redis && \
  cp -f *.conf /etc/redis

# Define mountable directories.
VOLUME ["/data"]

# Define working directory.
WORKDIR /data

# Expose ports.
EXPOSE 6379 
---------------------------------------------------------------------------------------------------------

in the log everything ok but there 
but at the end:

Updating module [default]...\Deleted [https://www.googleapis.com/compute/v1/projects/gssgcloudtest/zones/us-central1-f/instances/gae-builder-vm-20151201t161133].
Updating module [default]...failed.
ERROR: (gcloud.preview.app.deploy) Error Response: [13] Not enough VMs ready (0/1 ready, 1 still deploying). Version: 20151201t161133.388955482888900670
Georges-MacBook-Pro:mygss.ge Chipintoza$ 

i want to run Radis server for caching.
Can i run Radis on MCMs with support Sentinel ?

Thank you

**Comment by JustinBeckwith** [comment]
Throwing my hat in here.  You can in fact use memcache with nodejs on Managed VMs.  Check out this sample:
https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/master/appengine/express-memcached-session

No appengine-nodejs needed, just use the standard memcache module :) 

Now on how to do redis - please do not try to install Redis directly onto a Managed VM.  Here is a guide on how to set up redis (if you want to use this instead of the built in memcache) with App Engine:
https://cloud.google.com/nodejs/resources/databases/redis

I suggest using the redis labs option, at least initially.  

Next up the infamous "not enough VMs ready" error.  Sorry about that error - we're actively working on making that better.  To debug that one - you need to go to the developers console, and go to the cloud logging UI.  Look in the list of logs available, and look for a "crash.log".  This will have the console output from trying to start your node process.  

So TL;DR:
- I'd switch back to runtime:nodejs, move away from the customer dockerfile
- Use memcache if you can, it's built in and there's a sample!
- If you want to use redis, use redislabs or redisToGo

Hope this helps!

**Comment by Chipintoza** [comment]
@jonparrott  meny thanks to you... :) :8ball:

**Comment by Chipintoza** [comment]
@jonparrott  i make issues:
https://code.google.com/p/googleappengine/issues/detail?id=12583&thanks=12583&ts=1449143610

can you suggest something?

**Comment by jonparrott** [comment]
@Chipintoza you'll need to ensure you have `skip_files`in your app.yaml. See [here](https://github.com/GoogleCloudPlatform/nodejs-getting-started/blob/master/1-hello-world/app.yaml#L19).

**Comment by Chipintoza** [comment]
Yes i do this  i add  - ^node_modules
and now just works only:
runtime: custom
I generate Dockerfile and i use:
gcloud preview app deploy app.yaml --project gssgcloudtest --bucket gs://gss-staging-deploy

but every time i need to delete and create like this bucket gs://gss-staging-deploy

because i delete bucket staging.gssgcloudtest.appspot.com
gcloud alway generate error:
Synchronizing files to [gs://staging.gssgcloudtest.appspot.com/].
ERROR: (gcloud.preview.app.deploy) Could not synchronize files. The gsutil command exited with status [1]. Command output is 

I hope it will be corrected soon... ?

**Comment by jonparrott** [comment]
You shouldn't need to delete or create any buckets to deploy. Can you try with a new project?

**Comment by Chipintoza** [comment]
:( becous i delete staging.gssgcloudtest.appspot.com I have to do this :(...
later i will try with new project but it will be greate if will corrected.
thanks.
